### PR TITLE
Numerous changes related to ICAFIX, multi-run FIX, and the downstream scripts using those outputs

### PR DESCRIPTION
**Pull request by mharms**
This is a continuation of PR #83, following reversion of that PR after 
bc90cf1.  It has grown considerably in scope from the original name of PR #83 ("minor tweaks to FIX-related stuff"). Overall, it now includes:

1. Many improvements to code readability and consistency across scripts.
2. More consistent file naming scheme as regards variance normalization.
3. Addition of a polynomial detrend high-pass option to multi-run FIX (not supported yet in single-run FIX).
4. Deletion of various files to save space.

### CODE REVIEW

**Comment by mharms** [comment]
I need to do another pass through `ReApplyFixMultiRun.sh`.  It is a major pain to support scripts that are derived by cutting and pasting from a different upstream script.  We need to get away from that to the greatest extent possible. (e.g., Ideally, `hcp_fix_multi_run` and `ReApplyFixMultiRun.sh` would have been crafted initially as a single script with all the necessary functionality).

Meanwhile, some other questions have arisen while I was jumping around various scripts:

1. What is the role of `PostFix/DetectNonMatching.m`?
2. Any reason that `ReApplyFixMultiRun.sh` needs its own directory?  Can we move it into the ReApplyFix directory? (Analogous to how `hcp_fix` and `hcp_fix_multi_run` both reside in ICAFIX).
3. Why are we compiling our own version of fix_3_clean in `ReApplyFix/scripts`, rather than writing the scripts to use the compiled version that comes with the FIX distribution itself?  If we are going to do this, at a minimum, there should be a readme indicating what version of FIX it was compiled from, and also the MCR that it was compiled under (similar to what we have in the PostFix directory). Given that the variance normalization outputs of fix_3_clean weren't added until fix1.067, I suspect that our compiled version doesn't support that feature.

**Comment by glasserm** [comment]
1. That is a part of the PostReApplyFix script for FIX retraining.  Not sure why it is in there.
2. All the FIX related stuff could be in one directory as far as I care, but moving things around would require making sure nothing breaks.
3.  We shouldn't be doing that.

**Comment by glasserm** [comment]
I agree it is not optimal to have the separate scripts; however, at the time I didn't forsee the issues that have happened.

**Comment by mharms** [comment]
How aggressive to you want to get in putting all the FIX related stuff under one directory.  Should we move the contents of both PostFix and ReApplyFix to ICAFIX?  That seems logical to me.

Other than DeDriftAndResample, can you think of anything else that would have referred to either the PostFix or ReApplyFix directories?

**Comment by glasserm** [comment]
I think that is okay and the only stuff would be some launcher scripts of mine.  We should be sure any Example scripts are in line with these changes.

**Comment by glasserm** [comment]
I'm also a little confused which commits I need to review that are new from the prior pull request.

**Comment by coalsont** [comment]
1587ab310d3061fa721b0fab51fe467edde69f26 is the first commit that wasn't in the previous PR.

**Comment by glasserm** [comment]
I have looked over these commits and they seem okay, but we will need to test them.  I think both ReApply scripts need testing and also hcp_fix I think hcp_fix had some changes.  I don't recall anything but comments were changed in hcp_fix_multi_run, but Mike you might know different if there is anything that requires testing in there.

**Comment by mharms** [comment]
I have a couple other things I want to clean up before we test.

**Comment by mharms** [comment]
@xjqian Patches are needed from Oxford for several FSL binaries to support the use of large files (> 2GB).  We are hoping that FSL 6.0.1 will be ready for release within a couple weeks.  Either you must have those patches (either from us or Oxford), or your testing to date hasn't exceeded 2GB in the concatenated files?

**Comment by glasserm** [comment]
These look okay, are you ready for us to test?

**Comment by mharms** [comment]
I think these are about ready to test.  There are a couple small remaining issues that I'm aware of:
1) Need to a create a consolidated README file.  (I'll do that).
2) The issue remains as to whether the ReApply scripts should be using a version of fix_3_clean that we compiled (when --matlab-run-mode=0), or whether they should be using the one compiled by the FIX distribution.  Not sure if @coalsont or @tbbrown was responsible for that decision originally...
3) From what I can tell, neither of the ReApply scripts are logging the output of the matlab calls for interpreted matlab/octave.  Would that be accomplished by simply adding `" >> ${matlab_logfile} 2>&1 "` after the call to fix_3_clean within the `M_PROG` blocks?
4) Another Q for @coalsont or @tbbrown: Why aren't we using the 'call_matlab' script to "automatically" handle all the matlab mode stuff?  Did we feel it was just worthwhile to make the different matlab modes more transparent?
5) @glasserm:  Take a look in particular at fb5c406.  The previous conditional was wrong.  Is it supposed to be conditioned on DoVol instead?

**Comment by glasserm** [comment]
I'd like Tim to look at the question you directed at me as I think he wrote that bit, but otherwise looks okay.

**Comment by mharms** [comment]
Does the matlab/octave output get captured otherwise just to the usual stdout and stderr when in interpreted mode?  We should either have a convention in which all matlab/octave output gets captured into a separate log file, or it goes to stdout/stderr (with the possible exception of cases where we invoke using call_matlab.sh -- e.g., fix itself).  We definitely shouldn't be logging the output to different places depending on whether it is compiled or interpreted matlab mode (which is currently the case, if I'm understanding correctly).

@coalsont  So, is fb5c406 fine?

**Comment by glasserm** [comment]
For all my normal code it goes to stdout/stderr, but I don't know about these fancy call matlab scripts.

**Comment by coalsont** [comment]
fb5c406 looks like an improvement, but looking at it closer, maybe it should also be moved into in the conditional that chooses whether to run the matlab part or not (demeaning something that has already been demeaned should be harmless, I suppose).  The conditional on the matlab part looks like if the user deletes things and runs it with a surface registration name, it will rerun the matlab part without generating its volume outputs, not sure if we need to worry about that.

**Comment by coalsont** [comment]
It looks like the compiled version of reapply's fix_3_clean does capture to a log file only when using compiled.  I think I left it that way to not change the existing behavior, which was probably not the best choice.

**Comment by glasserm** [comment]
The 3 new commits look fine.

**Comment by mharms** [comment]
I'm trying to make sense of the various checks for existing files within ReApplyFixMultiRun.  From what I can tell, the purpose of the first 673 lines of code is to derive ${concatfmri}_hp${hp}.nii.gz and ${concatfmri}_Atlas${RegString}_hp${hp}.dseries.nii.  So, shouldn't we (couldn't we) have check for those files, and skip to the fix_3_clean stage if they exist?  We kind of have something similar to that, in that we currently have conditionals that check whether those two files exist before proceeding with all the merging code. But it seems like that check should actually be occurring before the initial `for fmriname in $fmris` loop.

**Comment by glasserm** [comment]
That is correct.

**Comment by mharms** [comment]
What do you think of renaming hcp_fix to ApplyFix.sh and hcp_fix_multi_run to ApplyFixMultiRun.sh, and dropping "Pipeline" from the two "ReApply*" scripts?  Also, could we maybe have a different name for ApplyHandReClassifications.sh?  (I want to avoid any implication that that script actually "applies" in terms of FIX).  Maybe "ReclassifySignalNoise.sh" ?

**Comment by glasserm** [comment]
I think renaming a bunch of files will be annoying.

**Comment by mharms** [comment]
Let me argue for renaming a little more:
1) ApplyFix.sh and ApplyFixMultiRun.sh would then have a very nice naming symmetry with the ReApply* scripts.
2) Avoiding 'hcp_fix' is good, since that is also the name of what is now a rather different version in the FIX distribution.
3) MR-FIX hasn't really been used externally yet, so I think we are ok with a name change there.  'hcp_fix' has certainly been used, but I think renaming that to ApplyFix.sh would on balance still be advantageous, per (1) and (2) above.

**Comment by glasserm** [comment]
If we are going to be disruptive like that, merging the functionality into two or even one script would make more sense.  I don't want to just rename scripts and have to deal with that (and all the folks who do use it have to deal with that).

**Comment by mharms** [comment]
Is there any particular reason that we need to use bash redirection for launching the interpreted matlab/octave modes?
i.e., 
rather than
```
(source "${FSL_FIXDIR}/settings.sh"; "${interpreter[@]}" <<M_PROG
${ML_PATHS} fix_3_clean('${fixlist}',${aggressive},${MotionRegression},${AlreadyHP});
M_PROG
)
```
could we just do
```
matlab_cmd="${ML_PATHS} fix_3_clean('${fixlist}',${aggressive},${MotionRegression},${AlreadyHP});"
(source "${FSL_FIXDIR}/settings.sh"; "${interpreter[@]}" "${matlab_cmd}")
```

**Comment by coalsont** [comment]
@mharms: your form does not give the script to standard input, but to the command line arguments.  It is not clear whether that would require a different invocation of matlab/octave.  A pipe is the most common construct for providing standard input, but a "here-document" may be preferred for style reasons.

**Comment by coalsont** [comment]
FYI, MRFIX seems to still run to completion as of commit 9efe83c.

**Comment by xjqian** [comment]
@mharms indeed, we haven't yet run into cases of concatenated file size larger than 2GB in our local testing.

**Comment by glasserm** [comment]
Are we done with changes?

**Comment by mharms** [comment]
I think we are getting close.  I've started doing some testing, and was hoping that I'd be able to run hcp_fix_multi_run using interpreted Octave and just two 5 min runs.  But Octave crashed at the 2nd pass call to functionhighpassandvariancenormalize:
`sh: line 1:  2876 Killed                  /export/octave-4.2.1/bin/octave-cli -q --no-window-system --eval "addpath('/scratch/mharms/fix1.067'); addpath('/export/fsl-6.0.0_OpenBLAS//etc/matlab'); addpath('/home/mharms/GitHub/HCPpipelines/ICAFIX/scripts'); functionhighpassandvariancenormalize(0.800000 , -1, 'rfMRI_2_4_hppd2', '/export/src/workbench-rh_linux64-v1.3.2/workbench/bin_rh_linux64/wb_command');" >> .fix.functionhighpassandvariancenormalize.log 2>&1`

Any insight to why that wouldn't work?  I'm going to try again with the most recent commits, but I don't think this octave crash will be fixed by any of my recent changes.

On a different note, I noticed that stdout when running PostFix.sh with Octave has a lot of blank lines corresponding to the middle of prepareICAs.m.  Any idea of what is going on there?

@tbbrown  I'm waiting to see if you have any comments about why we went the route of compiling our own version of fix_3_clean for the ReApply*.sh scripts (https://github.com/Washington-University/HCPpipelines/pull/87#issuecomment-451533503)

**Comment by coalsont** [comment]
Note that you will need a custom build of octave to use it on large concatenated files.  See https://github.com/coalsont/GNU-Octave-enable-64 for how I do this in my container (fork of someone else's work, I just wanted to make sure it didn't suddenly change on me).

Also, why is it running in sh?  I actually expect "Killed" to mean something else killed it (job monitoring system, out of memory, etc), while if it is octave's fault, I would expect an error more appropriate to the problem (invalid malloc size, segfault, etc).

I saw the blank lines, but didn't investigate.  If a multiline script is being fed into stdin, maybe it is half-displaying the interactive prompt?

**Comment by mharms** [comment]
Do we want to eliminate the use of the .*.log files within hcp_fix and hcp_fix_multi_run?  My reading of call_matlab.sh is that this should just a matter of removing the '-l' argument from call_matlab, in which case stdout and stderr for the matlab call continue to be whatever they were.

**Comment by glasserm** [comment]
I would be fine with that.

**Comment by mharms** [comment]
If someone runs FIX with two different high pass filters we currently have a naming conflict with the PostFix scene files, and the ReClassifyAs*txt files.  Our options are to either:
1) Add the hp string to their file names within ${fMRIName} directory, or 
2) Move them to reside within the ica directory.

Which one do you prefer?  (I personally favor (2)).

**Comment by glasserm** [comment]
That is not a supported configuration.  I don't really want to support it.

**Comment by mharms** [comment]
Why not?  There really isn't anything invalid about wanting to try two different highpass filters.  If we don't fix this, someone will do it, and then the conflicting file names will lead to problems/mistakes (without the user potentially even realizing that there is a conflict).

**Comment by glasserm** [comment]
Because it will be disruptive to existing code.

**Comment by mharms** [comment]
@coalsont  What is -g option of declare supposed to do in ApplyHandReClassifications?  I can't find any information on it, and it doesn't appear to be supported in the version of bash on the CHPC:

```
/home/mharms/GitHub/HCPpipelines/ICAFIX/ApplyHandReClassifications.sh: line 155: declare: -g: invalid option
declare: usage: declare [-aAfFilrtux] [-p] [name[=value] ...]
```
Can we do 
`declare -i "${2}"["${component}"]=1`
as the command instead?

**Comment by mharms** [comment]
@glasserm  Regarding the file name conflict: This is a case where some disruption is unfortunately necessary to avoid users getting burned by insidious little bugs.  I really think it would be a mistake not to fix that, because we should be generating robust scripts.

**Comment by glasserm** [comment]
We could make the change once that whole codebase is in gitrepo and you can make the necessary changes.  Basically that file gets used by all of the FIX retraining code too.

**Comment by mharms** [comment]
That's what I was going to do.  We have PostFix, ApplyHandReClassification, and the ReApply* scripts that will need changes to the relevant file paths and or file names?  Any others?

I was going to make the change now, since I'm actually testing those scripts...

**Comment by glasserm** [comment]
Obviously there are more scripts or I wouldn't have said it was a problem.

**Comment by mharms** [comment]
How about we make the change now to what is actually in git, and then I'll change the other retraining-related scripts at whatever point in the future they get added to the repo?

**Comment by glasserm** [comment]
This also would cause backwards incompatibility with existing results.  I do not think this is a helpful change.  If you like you can make it an error to run PostFix without deleting the old files.

**Comment by mharms** [comment]
How is this for a compromise: What if the actual files go into the .ica directory, but we create symlinks with the same name as the current file names within the ${fMRIName} directory?  Then, we have a permanent copy in the .ica directory, but things continue to work as they currently do.

**Comment by coalsont** [comment]
@mharms the -g option forces the variable to be declared in global scope, as otherwise (as I understand it, at least in recent bash) declare is implicitly local when used in a function.  The purpose of the line is to put the output into a global array variable (with name specified by an argument to the function), as bash functions can't "return" anything complicated the way other languages do.

It may be best to just repeat a pattern of code for loading the files rather than using this trick to do indirection of variable names.  Alternatively, it could have a hardcoded output name which gets copied into the desired name after the function call, as that would remove the need for the declare trick.

**Comment by glasserm** [comment]
If Tim thinks that will work I am okay with it.

**Comment by coalsont** [comment]
I don't have any insight as to the filename conflict issue.

**Comment by mharms** [comment]
I don't see a need to use `declare` at all.
`eval "${2}[${component}]=1"` within the `list_file_to_lookup()` function appears to work perfectly fine.

**Comment by coalsont** [comment]
I have removed the indirection, and after some puzzling about of bash's array quirks, have changed to a slightly less tricky solution.  I have pushed the change to my fork's master at 73c419df3e170e5797878ed10c72ed45ada0e5f0, reconciling it with mike's PR will be a little tricky since his PR is on his master (this is why people usually make a branch in order to propose changes, so they can keep their master tracking upstream, so they can pull in upstream changes easily).

If I push this to the upstream master, mike can resync by making a new branch with all his progress (this PR), rewinding his master to f65577c6cc80fe6ba2a53cb2213b3ce6d76c5da5, pulling the new upstream master, rebasing his in-progress branch, setting his master to the tip of that branch, and then force-pushing to his fork.  This would result in the cleanest history graph, and not worrying about what happens when merging two branches that both contain the same change.  If mike commits all his in-progress changes to his master, I could do this (thanks to the PR letting me force-push his master), and then he would either need a fresh clone of his fork, or to manually reset his local master.

Alternatively, I could push this as a commit onto this PR, but then users won't get this change until we merge this PR.

**Comment by coalsont** [comment]
I avoid eval as bad practice, as it re-parses whatever string results from the rest of the line, such that you end up with 2 bash parsing stages in a single command.  It would be safe in this case, as we have control of all the values being used (in other cases, if someone has any shell-special character in a filename, eval can do something very bad), but I think it is better to remove the indirection.

**Comment by mharms** [comment]
Testing ongoing.  I'll update this afternoon. @coalsont At this point, given that I revamped the style of ApplyHandReclassifications.sh, I think it'll be easiest if I just incorporate your changes.  Up to those changes, there were no merge conflicts.

**Comment by mharms** [comment]
We should have a consistent default for the matlab mode.  Currently PostFix defaults to compiled, while the ReApply* scripts to interpreted matlab.  Which do we want the default to be?

**Comment by mharms** [comment]
Is there any reason to keep the "fake" NIFTI Atlas.nii.gz file generated by fix_3_clean (esp. given that we have that exact same info as a proper CIFTI in the Atlas_hp_preclean.dtseries.nii file?

**Comment by glasserm** [comment]
1. Default to interpreted matlab.
2. No need for fake NIFTI.
3. Will we continue to make links to the ResultsFolder for the PostFix files?

**Comment by mharms** [comment]
1 and 2. Ok. (Will add that tomorrow).  3. Yes, per f81d71e, there are now symlinks in ResultsFolder for the files created in PostFix.  The physical scene and ReclassifyAs*txt files now reside in FIXFolder. 

@coalsont  7b9305d  implements your solution to the 'declare -g' problem, but in the context of my other harmonizations from 80bf8e9.  If you update your master to 7b9305d then we should get back to a situation with no merge conflicts.

More broadly things are looking good.  Here is what I've all successfully tested over the last 2 days:

**hcp_fix**

- hp=2000, absolute paths, no training data file specified, compiled matlab (in fix1.067 distribution)
- hp=2000, relative paths, training data file specified, compiled matlab (in fix1.067 distribution)
Both of the preceding used motion parameter regression
- hp=2000, relative paths, no training data data, Octave

**hcp_fix_multi_run** (just 2 files of 100 time points each)
- hp=pd2, relative paths, motion parameter regression

**PostFix**
- single run, compiled matlab (compiled binaries provided in the repo)
- single run, Octave
- multi-run (i.e., on output of hcp_fix_multi_run), compiled matlab

**ApplyHandReClassifications**
- single run, hp=2000
- multi-run, hp=pd2

**ReApplyFix**
- single run, with "Hand" edits applied, motion regression, compiled matlab (fix_3_clean provided in our repo, which does NOT include the _vn file outputs)
- single run, with --reg-name=MSMAll (i.e., doVol=0), using Octave

I'm testing ReApplyFixMultiRun now.

So, other than possibly ReApplyFixMultiRun, I think all the kinks have been worked out.

Remaining issues are:

1. Create integrated README.
2. Whether to eliminate use of the .log files in hcp_fix and hcp_fix_multi_run
3. What to do with the version of fix_3_clean that we have compiled as part of the repo.

**Comment by glasserm** [comment]
2.  We can stop using these.
3.  Delete it if it is out of date.

**Comment by mharms** [comment]
One comment about rerouting the log files to the regular stdout and stderr: functionhighpassandvariancenormalize generates a lot of output.  Is all that output necessary?  Could it be trimmed back?

**Comment by glasserm** [comment]
That info does help you make sure that the code is converging properly.

**Comment by mharms** [comment]
@glasserm  Please check b353514.  If the movement regressors were supposed to get re-computed in ReApplyFixMultiRunPipeline.sh, that wasn't happening.
Also, in the process of thinking about what can be skipped in ReApplyFixMultiRun (we already have conditionals that skip various stages if files already exist), I got to wondering about the following: Are the outputs from functionhighpassandvariancenormalize (i.e., "1st pass VN") dependent in any way on the FIX classification? -- I assume not, but want to confirm.

**Comment by glasserm** [comment]
Movement regressors shouldn't ever need to be recomputed.  

VN is not dependent on FIX classification, but FIX classification could be dependent on VN.  

The appropriate files need to be put in place for a new ${RegName}, but should exist for a recleaning with new classification.

File: DeDriftAndResample/DeDriftAndResamplePipeline.sh [hunk 1]
```diff
@@ -199,7 +199,7 @@ get_options()
 				p_SmoothingFWHM=${argument#*=}
 				index=$(( index + 1 ))
 				;;
-			--highpass=* | --high-pass=*)
```
[the remark this answers is not in the document]
Comment (mharms) [reply DeDriftAndResample/DeDriftAndResamplePipeline.sh]: Will be resolved in my next push.

**Comment by mharms** [comment]
Ok.  Other than needing to create an integrated README, from my perspective, this is done.
I've tested all the scripts within ICAFIX, but @coalsont please test the MR-FIX stuff within the context of your container.

**Comment by coalsont** [comment]
The MR FIX script ran to completion at commit 42479fc.  I will start another from the current PR state.

**Comment by glasserm** [comment]
What about the conflicts?

**Comment by mharms** [comment]
I proposed above that Tim simply update the master with the versions in my PR, and then the conflict goes away.

**Comment by coalsont** [comment]
Some notes on git generally accepted practices - "just use my version" is like a military solution to a diplomatic problem.  You are not intended to manually make duplicate edits as someone else, as this will always create conflicts if a merge is later attempted.  It is generally not advised to let any PR become a "kitchen sink", as it increases the chance of someone else editing the same part of the same file before the PR is ready to merge.

If @mharms is done making edits, I can resolve the conflicts while leaving the PR open by merging upstream into it and pushing to his branch.

**Comment by coalsont** [comment]
Also, commit 837afb921b0cb107054908299e1b828c93ee3c21 ran MR FIX to completion.

**Comment by mharms** [comment]
However you want to resolve this is fine with me.  Next time I'll create a branch at the start.  Obviously, I never envisioned that this PR would grow so large when I started it!  I'm working on the README now (if you want to wait for that to be done...)

**Comment by mharms** [comment]
One (hopefully) last thing: Should I move the contents of RestingStateStats into ICAFIX, since that is where it naturally belongs.  (We'll need to get Matt's dev version incorporated into the master, but moving RestingStateStats into ICAFIX will at least set expectations correctly as to where to find it going forward).

**Comment by glasserm** [comment]
No, it is a separate pipeline and will also interface with temporal ICA which will have its own folder.

**Comment by mharms** [comment]
In that case, then to my knowledge, this PR is done.

**Comment by coalsont** [comment]
I haven't done the full testing yet, I wanted to resolve the conflicts first (I moved the "set -e" back to the top of the ApplyHand... file, it is unfriendly to bury it in the middle of the script, and the indentation had become mixed spaces and tabs).  Other things have kept me busy, I'll try to get the testing done before going on vacation next week.

**Comment by glasserm** [comment]
Tim's tests look good aside from the issue with the moved PostFix scenes not working.  This needs to be fixed and then we can merge.

If it is VERY important that we keep the files moved, I suggest leaving the text files as they are and moving the scenes back to their original location, but adding the _${HighPass} string to their names to distinguish them.

**Comment by mharms** [comment]
I've returned the scene files to the ResultsFolder, but added `_hp${HighPass}` to their file name. Short of forcefully deleting any existing .ica folder to guarantee no naming conflicts (which I think is an unnecessary and extreme solution), I think this is the best overall approach.

**Comment by glasserm** [comment]
Did you test it?  Tim is out this week... 

If we have confirmed this is working I can merge.

**Comment by mharms** [comment]
Yes.  I ran it to completion and successfully opened the resulting single and dualscreen .scene files.

**Comment by mharms** [comment]
I'd reference the PR -- the opening of PR #87 is still a pretty good overall summary.  Maybe:
"Merge PR #87: (1) Adds polynomial detrend option to multi-run FIX; (2) Consolidation of all the FIX-related scripts into a single directory; (3) Many improvements to code readability and consistency"

**Pull request merged**